### PR TITLE
feat(vue-app): build indicator

### PR DESCRIPTION
**Pull request by pi0**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Show build progress inside pages!

nuxt-hmr-indicator

## TODO
  - [x] improve style
  - [x] support custom base URL
  - [x] add an option to optionally disable
  - [x] wait little more to display progress of 100

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. (PR: https://github.com/nuxt/docs/pull/1421)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by Atinux** [comment]
Awesome work Pooya!
Will do some small changes for the UI and will be good for me :)

File: packages/vue-app/template/components/nuxt-hmr.vue [hunk 1]
```diff
@@ -0,0 +1,103 @@
+<template>
+<transition appear>
+  <div class="nuxt__hmr" v-if="building">
+      <div>Building ({{ progress}}%)</div>
```
Comment (clarkdo) [inline packages/vue-app/template/components/nuxt-hmr.vue]: `space`

File: packages/vue-app/template/components/nuxt-hmr.vue [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        }
+      } catch (e) {}
+
+      this.building = !data.allDone
+      this.progress = Math.round(data.states.reduce((p, s) => p + s.progress, 0) / data.states.length)
+    },
+
+    wsClose() {
+        if (this.ws) {
+        this.ws.close()
```
Comment (clarkdo) [inline packages/vue-app/template/components/nuxt-hmr.vue]: indentation

File: packages/vue-app/template/components/nuxt-hmr.vue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    },
+
+    onWSMessage(msg) {
+      let data = msg.data
+
+      try {
+        if (data[0] === '{') {
+          data = JSON.parse(data)
+        }
+      } catch (e) {}
```
Comment (clarkdo) [inline packages/vue-app/template/components/nuxt-hmr.vue]: If data is not json, is following code supposed to still continue executing?

The `allDone` and `states` are all `undefined` which will cause exception as well.

File: packages/vue-app/template/components/nuxt.js [hunk 4]
```diff
@@ -12,12 +12,15 @@ import NuxtError from './nuxt-error.vue'
 <% } %>
 import NuxtChild from './nuxt-child'
 
+<%if (isDev) { %>import NuxtHMR from './nuxt-hmr'<% } %>
```
Comment (clarkdo) [inline packages/vue-app/template/components/nuxt.js]: Also wrapped with `client` ?

File: packages/vue-app/template/components/nuxt-hmr.vue [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+      this.ws = new WebSocket(this.wsURL)
+      this.ws.onclose = this.onWSClose.bind(this)
+      this.ws.onerror = this.onWSError.bind(this)
+      this.ws.onmessage = this.onWSMessage.bind(this)
+    },
+
+    wsReconnect(e) {
+      setTimeout(() => {
+        this.wsConnect()
```
Comment (clarkdo) [inline packages/vue-app/template/components/nuxt-hmr.vue]: Will this lead to infinite re-connect ?

File: packages/vue-app/template/components/nuxt-hmr.vue (hunk 1 above)
Comment (pi0) [reply packages/vue-app/template/components/nuxt-hmr.vue]: fixed

File: packages/vue-app/template/components/nuxt-hmr.vue (hunk 5 above)
Comment (pi0) [reply packages/vue-app/template/components/nuxt-hmr.vue]: i'll add a limit of 10

File: packages/vue-app/template/components/nuxt.js (hunk 4 above)
Comment (pi0) [reply packages/vue-app/template/components/nuxt.js]: This component is SSR safe and renders an empty node (only for dev)

**Comment by Atinux** [comment]
My last commit improves the style and uses `$nextTick` to show the 100% everytime.

nuxt-hmr-indicator

**Comment by Atinux** [comment]
About the option to disable it, what about `build.indicator: false`?

File: packages/config/types/build.d.ts [hunk 6]
```diff
@@ -28,6 +28,7 @@ interface NuxtBabelOptions extends Pick<TransformOptions, Exclude<keyof Transfor
 
 export interface NuxtConfigurationBuild {
   analyze?: BundleAnalyzerPlugin.Options | boolean
+  indicator?: boolean
```
Comment (kevinmarrec) [inline packages/config/types/build.d.ts]: Move `indicator` below (between `html` & `loaders`) so it fits alphabetical consistency

**Review by kevinmarrec — DISMISSED** [review]
Awesome work guys :) !

@Atinux I'm fine with `build.indicator`, unless we decide it needs more context.

File: packages/config/types/build.d.ts (hunk 6 above)
Comment (Atinux) [reply packages/config/types/build.d.ts]: Fixed, thanks Kevin, I don't know how I missed this!

**Comment by Atinux** [comment]
UI changes to show only Nuxt logo + percent:
nuxt-hmr-indicator

**Comment by pi0** [comment]
Updates:
- Added name to component and renamed to `buildIndicator`
- Smooth progress animation between updates

anim

**Comment by Atinux** [comment]
Final GIF

nuxt-hmr-indicator

**Review by atinux — APPROVED** [review]

**Review by kevinmarrec — APPROVED** [review]

**Pull request merged**